### PR TITLE
jsonStatAPI + terminal commands

### LINKED ISSUE #25: Terminal command responder
Terminal commands are being parsed but not actually handled correctly
Loading and unloading of plugins should be the main feature

### PR DESCRIPTION
**Pull request by maartenRamadan**
TODO terminal command shutdown

### CODE REVIEW

File: src/_includes/server/http/BodyBuilder.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#define BODYBUILDER_HPP
+
+#include <string>
+#include <map>
+#include <vector>
+
+namespace NotApache {
+
+	class BodyBuilder {
+	private:
```
Comment (mrjvs) [inline src/_includes/server/http/BodyBuilder.hpp]: Rename to JsonBodyBuilder

File: src/plugins/JsonStatAPI.cpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+bool JsonStatAPI::onBeforeFileServing(NotApache::HTTPClient& client) {
+	struct stat buf;
+	std::string body;
+
+	if (::stat(client.file.path.c_str(), &buf) == -1)
+		ERROR_THROW(NotFound());
+
+	std::string fileName = client.file.path.substr(client.file.path.rfind("/")+1);
+	std::string fileType = client.file.path.substr(client.file.path.rfind(".")+1);
```
Comment (mrjvs) [inline src/plugins/JsonStatAPI.cpp]: use `client.file.getExt()`

File: src/plugins/JsonStatAPI.cpp [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	client.data.response.builder.setBody(NotApache::BodyBuilder()
+		.line("name", fileName)
+		.line("type", fileType)
+		.line("size", buf.st_size)
+		.line("MIME type", MimeTypes::getMimeType(client.file.getExt()))
+		.time("last accessed", buf.st_atime)
+		.time("last modified", buf.st_mtime)
+		.time("last status change", buf.st_ctime)
+		.mode("mode", buf.st_mode)
```
Comment (mrjvs) [inline src/plugins/JsonStatAPI.cpp]: Use camelCase for naming

File: src/server/responder/serving.cpp [hunk 4]
```diff
@@ -86,6 +88,20 @@ void HTTPResponder::serveFile(HTTPClient &client) {
...
+			if (it->second && it->first->onBeforeFileServing(client)) {
+				client.data.response.setResponse(client.data.response.builder.build());
+				return;
+			}
+		}
+		catch (const std::exception& e) {
+			globalLogger.logItem(logger::ERROR, std::string("Json_stat_API error: ") + e.what());
+			handleError(client, 404);	
+		}
+	}
```
Comment (mrjvs) [inline src/server/responder/serving.cpp]: Move the hook to line 35 of this file. that will also include serving of index files

File: src/server/terminal/TerminalResponder.cpp [hunk 5]
```diff
@@ -20,5 +20,31 @@ void TerminalResponder::respond(const std::string &str) {
...
+					<< "\t\t- help" 				<< std::endl << std::endl
+					<< "\tload/unload plugins:" 	<< std::endl
+					<< "\t\t- load page_404_gen"	<< std::endl
+					<< "\t\t- unload page_404_gen"	<< std::endl
+					<< "\t\t- load json_stat_api" 	<< std::endl
+					<< "\t\t- unload json_stat_api" << std::endl;
+	}
+
 	globalLogger.logItem(logger::INFO, "Handled terminal command :)");
 }
```
Comment (mrjvs) [inline src/server/terminal/TerminalResponder.cpp:65]: Handle commands differently:
1. split on whitespace (args[])
2. compare args[0] with list of available commands. if not found send usage
3. for load/unload, find args[1] (the plugin), display error if not found and if already loaded/unloaded
4. make sure to send success

Comment (mrjvs) [reply src/server/terminal/TerminalResponder.cpp:65]: And for exit, pass a Server class pointer to the arguments of respond(); (which is `this` when you call it)
then just call `server->shutdown();` to exit

File: src/server/responder/serving.cpp [hunk 6]
```diff
@@ -86,6 +88,20 @@ void HTTPResponder::serveFile(HTTPClient &client) {
 	// if "*" set to first language in config
 	if (client.data.request.data.acceptLanguage == "*")
 		client.data.request.data.acceptLanguage = client.routeBlock->getAcceptLanguage()[0];
+	
+	// loops through plugins and executes if plugin is loaded
+	for (plugin::PluginContainer::pluginIterator it = globalPlugins.begin(); it != globalPlugins.end(); ++it) {
+		try {
+			if (it->second && it->first->onBeforeFileServing(client)) {
+				client.data.response.setResponse(client.data.response.builder.build());
```
Comment (mrjvs) [inline src/server/responder/serving.cpp]: move to hook code

File: src/plugins/JsonStatAPI.cpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+		.build()
+	);
+
```
Comment (mrjvs) [inline src/plugins/JsonStatAPI.cpp]: Set mimeType of `client.data.response.builder` to `application/json`

File: src/server/http/BodyBuilder.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+BodyBuilder& BodyBuilder::line(const std::string& type, size_t value) {
+	_body += _tab + typeFormat(type) + numberFormat(value);
+	_body += _endLine;
+	return *this;
+}
+
+const std::string& BodyBuilder::build() {
+	_body += "}";
+	_body += _endLine;
+	return _body;
```
Comment (mrjvs) [inline src/server/http/BodyBuilder.cpp]: Store a std::map<std::string,std::string> and build at the end

File: resources/example-configs/cgi.conf [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        index index.html
+        root    ../resources/site
+        directory_listing true
+        #cgi ../resources/site/index.html
+#         cgi_ext .htm
+    }
+}
+
+use_workers 42
+
```
Comment (mrjvs) [inline resources/example-configs/cgi.conf]: remove this file

**Review by mrjvs — CHANGES_REQUESTED** [review]

**Comment by mrjvs** [comment]
also strmode doesnt exist on linux. make your own implementation using [S_IS* macros](https://www.gnu.org/software/libc/manual/html_node/Testing-File-Type.html)

File: src/server/responder/serving.cpp [hunk 10]
```diff
@@ -30,6 +32,23 @@ void HTTPResponder::prepareFile(HTTPClient &client, const struct ::stat &buf, in
...
+			catch (const std::exception& e) {
+				globalLogger.logItem(logger::ERROR, std::string("Json_stat_API error: ") + e.what());
+				handleError(client, 404);	
+			}
+		}
+	}
+	// return so the plugin changes don't get overwritten
+	if (plugins)
+		return ;
+
```
Comment (mrjvs) [inline src/server/responder/serving.cpp]:
```suggestion
	for (plugin::PluginContainer::pluginIterator it = globalPlugins.begin(); it != globalPlugins.end(); ++it) {
		if (it->second && it->first->onFileServing(client)) {
			return;
		}
	}
```

File: src/server/terminal/TerminalResponder.cpp [hunk 11]
```diff
@@ -5,20 +5,58 @@
...
+	if (command[0] == "load") {
+		plugin::PluginContainer::pluginIterator it = globalPlugins.find(command[1]);
+		if (it != globalPlugins.end()) {
+			if (it->second == true)
+				globalLogger.logItem(logger::INFO, "Plugin already loaded");
+			else {
+				it->second = true;
+				globalLogger.logItem(logger::INFO, "Plugin loaded");
+			}
+		}
```
Comment (mrjvs) [inline src/server/terminal/TerminalResponder.cpp:37]: add plugin not found for this if branch

File: src/server/terminal/TerminalResponder.cpp [hunk 12]
```diff
@@ -5,20 +5,58 @@
...
-			it->second = false;
+	else if (command[0] == "exit")
+		server->shutdownServer();
+	else {
+		if (command[0] != "help")
+			globalLogger.logItem(logger::INFO, "Invalid command...");
+		globalLogger.logItem(logger::INFO, _help);
 	}
+
 	globalLogger.logItem(logger::INFO, "Handled terminal command :)");
```
Comment (mrjvs) [inline src/server/terminal/TerminalResponder.cpp]: you can remove this line now

File: src/_includes/plugins/JsonStatAPI.hpp [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+		virtual ~JsonStatAPI();
+
+		virtual bool onFileServing(NotApache::HTTPClient &client);
+
+		class NotFound : public std::exception {
+			public:
+				virtual const char* what() const throw() { return "file not found"; }
+		};
+	};
+	//TODO MAP VOOR KEYS om dubbele te voorkomen
```
Comment (mrjvs) [inline src/_includes/plugins/JsonStatAPI.hpp]: remove todo, already done

**Review by mrjvs — CHANGES_REQUESTED** [review]

**Comment by mrjvs** [comment]
also do all todos in code

**Review by mrjvs — CHANGES_REQUESTED** [review]
make it compile then all is good (you can check the compilation logs at the bottom of the pr)

**Review by mrjvs — APPROVED** [review]

**Review by pmerku — APPROVED** [review]

**Pull request merged**